src/array/cuda/labor_sampling.cu

```diff
@@ -0,0 +1,807 @@
+  const IdType * rows;
+  const IdType * indptr;
+  const FloatType * ds;
+  IdType * in_deg;
+  FloatType * cs;
+  __host__ __device__
+  auto operator() (IdType tIdx) {
+    const auto out_row = rows[tIdx];
+    const auto d = indptr[out_row + 1] - indptr[out_row];
+    in_deg[tIdx] = d;
+    cs[tIdx] = num_picks / (ds ? ds[tIdx] : (FloatType)d);
+  }
+};
+
+template <typename IdType, typename FloatType>
+__global__ void _CSRRowWiseOneHopExtractorKernel(
+  const uint64_t rand_seed,
+  const IdType hop_size,
+  const IdType * const rows,
+  const IdType * const indptr,
+  const IdType * const subindptr,
+  const IdType * const indices,
+  const IdType * const idx_coo,
+  const IdType * const nids,
+  const FloatType * const A,
+  FloatType * const rands,
+  IdType * const hop,
+  FloatType * const A_l) {
+  IdType tx = static_cast<IdType>(blockIdx.x) * blockDim.x + threadIdx.x;
+  const int stride_x = gridDim.x * blockDim.x;
+
+  curandStatePhilox4_32_10_t rng;
+  if (!labor)  // reduces to neighbor sampling, see r_t vs r_{ts} in arXiv:2210.13339
+    curand_init(rand_seed, tx, 0, &rng);
+
+  while (tx < hop_size) {
+    IdType rpos = idx_coo[tx];
+    IdType rofs = tx - subindptr[rpos];
+    const IdType row = rows[rpos];
+    const auto in_idx = indptr[row] + rofs;
+    const auto u = indices[in_idx];
+    hop[tx] = u;
+    const auto v = nids ? nids[u] : u;
+    if (labor)  // 123123 is just a number with no significance.
+      curand_init(123123, rand_seed, v, &rng);
+    const float rnd = curand_uniform(&rng);
+    if (A)
+      A_l[tx] = A[in_idx];
+    rands[tx] = (FloatType)rnd;
+    tx += stride_x;
+  }
+}
+
+template <typename IdType, typename FloatType, int BLOCK_CTAS, int TILE_SIZE>
+__global__ void _CSRRowWiseLayerSampleDegreeKernel(
+  const IdType num_picks,
+  const IdType num_rows,
+  const IdType * const rows,
+  FloatType * const cs,
+  const FloatType * const ds,
+  const FloatType * const d2s,
+  const IdType * const indptr,
+  const FloatType * const probs,
+  const FloatType * const A,
+  const IdType * const subindptr) {
+  typedef cub::BlockReduce<FloatType, BLOCK_SIZE> BlockReduce;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  __shared__ FloatType var_1_bcast[BLOCK_CTAS];
+
+  // we assign one warp per row
+  assert(blockDim.x == CTA_SIZE);
+  assert(blockDim.y == BLOCK_CTAS);
+
+  IdType out_row = blockIdx.x * TILE_SIZE + threadIdx.y;
+  const auto last_row = min(static_cast<IdType>(blockIdx.x + 1) * TILE_SIZE, num_rows);
+
+  constexpr FloatType ONE = 1;
+
+  while (out_row < last_row) {
+    const auto row = rows[out_row];
+
+    const auto in_row_start = indptr[row];
+    const auto out_row_start = subindptr[out_row];
+
+    const IdType degree = indptr[row + 1] - in_row_start;
+
+    if (degree > 0) {
+      // stands for k in in arXiv:2210.13339, i.e. fanout
+      const auto k = min(num_picks, degree);
+      // slightly better than NS
+      const FloatType d_ = ds ? ds[row] : degree;
```

Comment:
The implementations are exactly the same. They would produce almost the exact same output if I had used the same PRNG for both. There is a slight logic difference between the CPU and CUDA implementations in terms of arrays used etc. but otherwise the code does the same thing. The CPU side ds array holds the sum of weights if there are weights, otherwise it holds the actual degree. The CUDA side doesn't have a valid ds array when there are no weights so there is a conditional there.